util-hadoop/src/main/java/com/google/cloud/hadoop/util/RefreshTokenAuth2Provider.java

```diff
@@ -0,0 +1,160 @@
+/*
+ * Copyright 2022 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.cloud.hadoop.util;
+
+import static com.google.api.client.util.Preconditions.checkNotNull;
+import static com.google.cloud.hadoop.util.HadoopCredentialsConfiguration.AUTH_CLIENT_ID_SUFFIX;
+import static com.google.cloud.hadoop.util.HadoopCredentialsConfiguration.AUTH_CLIENT_SECRET_SUFFIX;
+import static com.google.cloud.hadoop.util.HadoopCredentialsConfiguration.AUTH_REFRESH_TOKEN_SUFFIX;
+import static com.google.cloud.hadoop.util.HadoopCredentialsConfiguration.PROXY_ADDRESS_SUFFIX;
+import static com.google.cloud.hadoop.util.HadoopCredentialsConfiguration.PROXY_PASSWORD_SUFFIX;
+import static com.google.cloud.hadoop.util.HadoopCredentialsConfiguration.PROXY_USERNAME_SUFFIX;
+import static com.google.cloud.hadoop.util.HadoopCredentialsConfiguration.TOKEN_SERVER_URL_SUFFIX;
+import static com.google.cloud.hadoop.util.HadoopCredentialsConfiguration.getConfigKeyPrefixes;
+import static com.google.common.flogger.LazyArgs.lazy;
+
+import com.google.api.client.auth.oauth2.ClientParametersAuthentication;
+import com.google.api.client.auth.oauth2.RefreshTokenRequest;
+import com.google.api.client.auth.oauth2.TokenRequest;
+import com.google.api.client.auth.oauth2.TokenResponse;
+import com.google.api.client.http.GenericUrl;
+import com.google.api.client.http.HttpTransport;
+import com.google.api.client.json.JsonFactory;
+import com.google.api.client.json.gson.GsonFactory;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableList;
+import com.google.common.flogger.GoogleLogger;
+import java.io.IOException;
+import java.time.Instant;
+import java.util.List;
+import org.apache.hadoop.conf.Configuration;
+
+/**
+ * Retrieve an access token using the OAuth2 refresh token grant flow. See <a
+ * href="https://datatracker.ietf.org/doc/html/rfc6749#section-1.5">RFC 6749</a>.
```

Comment:
After a second look, I think that it will be great if we contribute this OAuth2 flow to the Google Auth library and just use it in GCS connector after that: https://github.com/googleapis/google-auth-library-java

This way it could be used much more widely than just in GCS connector.

WDYT? Do you mind to send PR there?

Reply:
Looking more closely to this lib, they got an interesting class for us:

```
UserCredentials
```


What we could do, is instead of going on the OAuth2 provider road, we could introduce a new type, called `USER_CREDENTIALS` 
and then just use the config we got to build a `UserCredentials` in `HadoopCredentialsConfiguration#getCredentialsInternal`.

`UserCredentials` seems to do pretty much what we did in `RefreshTokenAuth2Provider`

Note: only things we haven't introduced in our config that we may need, is `quotaProjectId`. Not sure what it is to be honest, at least not in this OAuth2 context. IMO we could set it to null, from the code it seems to be handling null quite well. In fact, does nothing else than adding `quotaProjectId` (if not null) into the headers of the requests and that's not an expected header in OAuth2, more a tracing value by the look of it.

Reply:
Ok so I gave it a try and seems it would work.

Once thing though: the UserCredentials is not supported the fact that you could get a new refresh token, as part of the token response.
Not the end of the world, as we can assume that Google OAuth2 provider is setup in such a way that it would not issue new refresh token. If one day that changes, I guess the OAuth2 lib would release a fix and you would only need to upgrade the lib.
Annoying bit is that if they do so ^, then your previous version of the hadoop connector will stop working for the authentication type USER_CREDENTIAL.

@medb Okay with that limitation?

Reply:
Yes! PR looks good with few minor comments.